### ISSUE TITLE
Document declarations and filters versioning

### ISSUE THREAD

**Issue by Ndpnt** [ambanum/CGUs#156]
We need to be able to regenerate versions from snapshots. As documents is aim to change over time (location or filters) we can't rely on the last version of the declaration to regenerate the version from an old snapshot. So we need a system to keep track of declaration changes, that's what we called **declarations and filters versioning**.

At this time, we see three solutions which have in common the following rules:
 - `history` is optional
 - the current valid declaration has no date and should be clearly identifiable
 - the `valid_until` date is an inclusive expiration date. It should be the exact authored date of the last snapshot commit for which the declaration is still valid.

## Option 1: Add an `history` field in service declaration

In `services/ASKfm.json`:
```
{
  "name": "ASKfm",
  "documents": {
    "Terms of Service": {
      "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
      "select": ".selection",
      "filter": [ "add" ]
      "history": [
        {
          "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
          "select": "body",
          "filter": [ "add" ]
          "valid_until": "2020-08-24T14:02:39Z"
        },
        {
          "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
          "select": "body",
          "valid_until": "2020-08-23T14:02:39Z"
        }
      ]
    }
  }
}
```

Note: When no historisation is needed the file may have no mention of history.

**Pros:**
- Everything is in the same file:
    - might prevent to forget to update existing history
    - might help user to know that history is a thing and encourage them to learn about it if they feel the need
    - no (pseudo-)hidden knowledge about history

**Cons:**
- Apparent complexity can discourage new contributors
- With time, the file can become huge

## Option 2: Add an `serviceId.history.json` file

In `services/ASKfm.json`:
```
{
  "name": "ASKfm",
  "documents": {
    "Terms of Service": {
      "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
      "select": ".selection",
      "filter": [ "add" ]
    }
  }
}
```

In `services/ASKfm.history.json`:
```
{
  "name": "ASKfm",
  "documents": {
    "Terms of Service": [
      {
        "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
        "select": "body",
        "filter": [ "add" ]
        "valid_until": "2020-08-24T14:02:39Z"
      },
      {
        "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
        "select": "body",
        "valid_until": "2020-08-23T14:02:39Z"
      }
    ]
  }
}
```

**Pros:**
 - Service declaration stay small and simple
 - History file is kept close to the service declaration so users might see them

**Cons:**
 - Make the discovery of history capacities less easy
 - Increase the probability of forgetting to update history file when making a change in the service discovery

## Option 3: Add an history service declaration file in `services/history` folder

In `services/ASKfm.json`:
```
{
  "name": "ASKfm",
  "documents": {
    "Terms of Service": {
      "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
      "select": ".selection",
      "filter": [ "add" ]
    }
  }
}
```

In `services/history/ASKfm.json`:
```
{
  "name": "ASKfm",
  "documents": {
    "Terms of Service": [
      {
        "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
        "select": "body",
        "filter": [ "add" ]
        "valid_until": "2020-08-24T14:02:39Z"
      },
      {
        "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
        "select": "body",
        "valid_until": "2020-08-23T14:02:39Z"
      }
    ]
  }
}
```

**Pros:**
 - Service declaration stay small and simple
 - All history updates are reserved to users with the knowledge that might work as gatekeepers

**Cons:**
 - All history updates are reserved to users with the knowledge that might work as gatekeepers :)
 - Need to rely on people with knowledge to keep the history

### Community

The choice might have implication on the community that will grow around the project.

*Option 1* shows everything to everyone, it might frightened some contributors with some apparent complexity (once there are history in the declaration file), but it might also encourage them to learn about it if they want or feel the need to. All contributors will share the same view and knowledge about the system. This might encourage collaboration between them to learn and improve together.

*Option 2* and *Option 3* hide the complexity of history management in separate files and only most adventurous contributors will find them by themselves. Contribution to those files will probably be done by specific contributors that will be taught to manage those file. Thus creating two different kind of contributors: those who will stay with the basic service declaration, not knowing that more complex options exist, and those who will have the knowledge of history management whose work might stay in the shadow or work as gatekeeper.

**Comment by Ndpnt** [comment]
Here is a first version of an RFC about "declarations and filters versioning", what do you think of it @LucasVerneyDGE @michielbdejong @MattiSG? 🙂

**Comment by michielbdejong** [comment]
I see the point of hiding complexity, smart! I think it's fine as long as there is a clear section about "Reprocessing historical snapshots" in the docs which mentions the role of the `services/history/` folder.

**Comment by LucasVerneyDGE** [comment]
Same opinion here, I think the cons of the first option (apparent complexity) might overcome the pros (making it obvious how to handle history) for new users.

I don't have strong opinion on option 2 vs option 3, but a slight preference for option 3 to avoid confusion with `filters` and others "live" objects. Having a dedicated history folder would also let us have history of filters in the same, dedicated, place.

**Comment by Ndpnt** [comment]
@MattiSG do you have an opinion about this?

**Comment by MattiSG** [comment]
First of all thank you very much and congrats @Ndpnt for the clarity and the quality of the suggested options! 🙇

I am unfavorable of option 1, as I feel it breaks the clean readability we reached with the current document declaration system. I really like how we have a series of verbs that can be read as chronological instructions, and introducing a declarative `history` property breaks it. It also makes it much harder to read the entirety of the declaration: if each document has its own history, the file will be huge and hard to digest.

I have no strong preference between options 2 and 3. Maybe a slight preference for option 2, which increases the discoverability of the history for contributors who might become more “adventurous” 😉

In both cases, I don't see the point of repeating the `name` property. It can only lead to rot and divergence with the one in the original file, which should always be authoritative in any case. Maybe that history file should only contain the document declarations, making it clear that all metadata is only taken from the main declaration?

I would thus suggest:

## Option 2A



In `services/ASKfm.json`, **called the “service declaration”**:

```
{
  "name": "ASKfm",
  "documents": {
    "Terms of Service": {
      "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
      "select": ".selection",
      "filter": [ "add" ]
    }
  }
}
```

In `services/ASKfm.history.json`, **called the “service history”**:

```
{
  "Terms of Service": [
    {
      "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
      "select": "body",
      "filter": [ "add" ]
      "valid_until": "2020-08-24T14:02:39Z"
    },
    {
      "fetch": "https://ask.fm/docs/terms_of_use/?lang=en",
      "select": "body",
      "valid_until": "2020-08-23T14:02:39Z"
    }
  ]
}
```

**Comment by Ndpnt** [comment]
Implemented in #198
